hbase-server/src/main/java/org/apache/hadoop/hbase/master/region/MasterRegion.java

```diff
@@ -143,17 +154,54 @@ private void shutdownWAL() {
     }
   }
 
+  /**
+   * Performs the mutation to the master region using UpdateMasterRegion update action.
+   * @param action Update region action.
+   * @throws IOException IO error that causes active master to abort.
+   */
   public void update(UpdateMasterRegion action) throws IOException {
-    try {
-      action.update(region);
-      flusherAndCompactor.onUpdate();
-    } catch (WALSyncTimeoutIOException e) {
-      LOG.error(HBaseMarkers.FATAL, "WAL sync timeout. Aborting server.");
-      server.abort("WAL sync timeout", e);
-      throw e;
+    for (int tries = 0; tries < maxRetriesForRegionUpdates; tries++) {
+      try {
+        // If the update is successful, return immediately.
+        action.update(region);
+        flusherAndCompactor.onUpdate();
+        return;
+      } catch (RegionTooBusyException e) {
+        // RegionTooBusyException is the type of IOException for which we can retry
+        // for few times before aborting the active master. The master region might
+        // have genuine case for delayed flushes and/or some procedure bug causing
+        // heavy pressure on the memstore.
```

Comment:
If RegionTooBusyException is caught, here can trigger `flusherAndCompactor.onUpdate();`

Reply:
You mean for `tries == (maxRetriesForRegionUpdates - 1)` condition? Otherwise, it will do it anyways for all retries as per the above loop.

Reply:
I mean like this:
```
 } catch (RegionTooBusyException e) {
   flusherAndCompactor.onUpdate();
   if (tries == (maxRetriesForRegionUpdates - 1)) {
     ***
   }
 }
```

The time interval hasn't been reached, but the `changesAfterLastFlush` threshold has been met, could happen, so when this exception caught. you need to trigger `flusherAndCompactor.onUpdate();`

Or otherwise, you need to switch the execution order
```
flusherAndCompactor.onUpdate(); // first
action.update(region); // after
```

Reply:
Makes sense!